[Feature] Add Playwright

esizer opened the pull request:
🤖 Resolves #8809 

## 👋 Introduction

This is an example of [Playwright](https://playwright.dev/) implementation.

## 🕵️ Details

> [!NOTE]
> Unsure if this will get merged but it demonstrates how we could effectively implement playwright. I could take it further if the team is interested.

First off, this was very enjoyable and imo, much easier to work with and understand than Cypress. I'm going tot take my time to try and explain what is going on here but if you are reviewing and have questions, feel free to reach out.

### Async/Await

One of my largest gripes with Cypress is that it tries hard to hide async functionality from you. Playwright is different in that it embraces it and feels far less like magic. Instead of just assuming a navigation is waited on, Playwright makes you add `await` before to be explicit. 

### Fixtures

REF: https://playwright.dev/docs/test-fixtures

One of the largest changes was the idea of fixtures. In Cypress, we were just using them as static json. In Playwright, they are classes that encapsulates the current environment and functionality. So, in our setup we create an `adminPage` and `applicantPage` fixture that has that user logged in and isolated to that fixture. This means, we no longer need to keep logging in and out to do different things. As well, since these are just classes, you can add methods for common functionality for that user/page and extend them!

```ts
test("something", async ({ adminPage, applicantPage }) => {
  // This user can go to admin
  await adminPage.page.goto("/admin");

  // This user cannot (no need to logout/login)
  await applicantPage.page.goto("/admin");
});
```

### Projects

REF: https://playwright.dev/docs/test-projects

I'm not sure if cypress has this concept but it is neat. When you run `npx playwright test` it runs all projects at once in parallel. So we are testing chromium, firefox and webkit all that once with no extra work.

### Debugging

I never had any luck debugging tests in cypress but found playwright just worked. If you run `npx playwright test --debug` it opens the ui and a debugger console allowing you to step through your tests like a normal debugger. This made finding issues with my tests much easier.

## 🧪 Testing

> [!NOTE]
> I only added a single script and opted to pass flags for now since this was more exploratory (sorry 😅 ). We can add easier to run commands for ui and debugging if we decide we want to move forward.

1. Install `npm run install`
2. Build `bpm run build`
3. Run tests in headless `npm run e2e:playwright`
4. Confirm it works
5. Try out the ui `npm run e2e:playwright -- -- --ui`
6. Confirm it works
7. Try out the debug mode `npm run e2e:playwright -- -- --debug`

## 📸 Screenshot

![Screenshot 2023-12-29 153319](https://github.com/GCTC-NTGC/gc-digital-talent/assets/4127998/8e390b99-beb0-4bf3-a2b8-0f928cbab12b)

brindasasi commented:
This is cool. Love the parallel test runs. 
Debug mode opens up stack of windows at a time and lots of runners.
Do you know how to run single test or single suit?

brindasasi commented:
seems like we can automate this in Azure pipeline  as well 🥳
https://techcommunity.microsoft.com/t5/azure-architecture-blog/automating-end-to-end-testing-with-playwright-and-azure/ba-p/3883704

esizer commented:
> Do you know how to run single test or single suit?

Yeah, there should be accordions and you can expand them and press the play button. 

![test-runner](https://github.com/GCTC-NTGC/gc-digital-talent/assets/4127998/82ffb095-d421-4b62-9c26-6fe313c87521)

brindasasi commented:
This looks good to me and my only concern is the effort for migration of all existing test cases into this. 
As Tristan said , probably we could start with new test cases run both and then slowly migrate & get rid of cypress. 

Also did you check how do we do feature flags in here? 

I would leave this here for other team members review & comments.

brindasasi commented on apps/playwright/playwright.config.ts:
```diff
@@ -0,0 +1,86 @@
+    //   use: { ...devices['iPhone 12'] },
+    // },
+
+    /* Test against branded browsers. */
```
So far Edge is our top most used browser. We should keep it.

petertgiles replied:
I'm guessing this is disabled by default because it is assumed that Edge will behave exactly the same as Chromium.  Reasonable assumption?

esizer replied:
Yeah, that is what I gathered as well.

esizer commented:
> Also did you check how do we do feature flags in here?

Not yet but that does seem important. I can't imagine we will have issues so maybe I can take a look today? It looks like they have some pretty good documentation on [how to modify responses.](https://playwright.dev/docs/network#modify-responses)

petertgiles commented on apps/playwright/utils/user.ts:
```diff
@@ -0,0 +1,58 @@
+  preferredLanguageForExam: "EN",
+
+  // optional
+  telephone: undefined,
```
I think this stuff is a hangover from when Cypress was in plain JS.  It'd probably be better to just import the User type from graphql and type this object.

esizer replied:
Oh, thanks! I was kind of just copy and pasting a lot of stuff from Cypress to this so that is good to know 👍

petertgiles commented on apps/playwright/playwright.config.ts:
```diff
@@ -0,0 +1,86 @@
+  /* Run tests in files in parallel */
+  fullyParallel: true,
+  /* Fail the build on CI if you accidentally left test.only in the source code. */
+  forbidOnly: !!process.env.CI,
```
You don't anticipate any trouble running this in a GH workflow, right?

esizer replied:
Yeah, so my intention was to do the minimum amount of work to see what people thought of the style of writing tests before exploring it further.

 It seems like the ones who have reviewed it are tentatively positive about it so we may be able to start filing in some gaps like the CI thing. Do you think it is worth exploring that now? This is obviously a requirement and they have some pretty good [documentation around it ](https://playwright.dev/docs/ci-intro)(one of the things I enjoyed about working with it).

petertgiles replied:
No, that's fine.  It was a reasonable scope limit.  As long as it's well documented, that's enough for me right now.

petertgiles commented on apps/playwright/playwright.config.ts:
Is there Axe integration?

esizer replied:
Yup! There is an [axe package](https://www.npmjs.com/package/@axe-core/playwright) for it and playwright does a good job of [documenting how to use it](https://playwright.dev/docs/accessibility-testing) 😄

petertgiles replied:
Nice!

petertgiles reviewed (commented):
So far I like what I see!  The fixtures system looks cool but it's going to take a bit of head-wrapping since we never touch classes in JS. 

I think I need to mess around with this a bit more on Linux.  Most (but not all!) the tests failed for me.
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/cee1b251-7ede-4444-affa-f0d37a5ae45b)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/ab232407-aaf1-4039-8a80-267f4e7f1d31)

esizer commented:
> I think I need to mess around with this a bit more on Linux. Most (but not all!) the tests failed for me.

Oh shoot, I should have mentioned in the testing steps that would may need to run `npx playwright install-deps`. I also ran into this issue on WSL and that fixed it for me 👍

esizer commented:
> Unfortunately it was still not able to find some packages. Bummer, it seems like Playwright is only supported on LTS versions of Ubuntu. We could probably run the CLI version in Docker but that negates the nice debugging UI, right?

oh darn, I forgot you were not on an LTS 😢 Yeah, we may need to run in docker then which may make the debugging non-functional. We could try to map the ports it uses 🤔 Though, I'm not super optimistic about that 👎

yonikid15 reviewed (commented):
Wow looks good! The code was easy to follow, and I'm excited to try it out myself. Reviewing the classes gave me a flashback to when I working with Java 😆.

I was able to run the tests headless, using the UI, and debug mode. However, some of the tests are failing on my end (still looking into it) 😢. I'll post screenshots just incase someone else experienced something similar.

Headless:
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/5da56d6f-fda2-46fe-8179-826348e1ebd1)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/02ac6cdf-cc09-4447-b057-88573b583cb7)

Debug mode:
![Screenshot 2024-01-12 140838](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/6541597e-ed3e-4db8-b22b-038a8baa9b35)

esizer commented:
> I was able to run the tests headless, using the UI, and debug mode. However, some of the tests are failing on my end (still looking into it) 😢. I'll post screenshots just incase someone else experienced something similar.

I wasn't able to reproduce that exact error but I put some stuff in to slow them down. I think they were running too fast so the elements were hidden behind the loading spinner or something. Also noticed the skill library test name was not accurate 😛 

[improve tests, correct names](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8917/commits/a298357f19958b444c333d5f1b83f61cdd8813e6)

petertgiles commented:
I tried to get it running in our docker environment to see if I could get it going on my workstation.  Unfortunately I just got timeouts.  :cry: 
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/953a0122-33cb-4850-9f70-b996e9274af4)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/24e39433-2494-4055-91f3-ccdd9791cb2f)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/704defef-c41c-4d1f-8286-0673e08a6914)

yonikid15 commented:
> I tried to get it running in our docker environment to see if I could get it going on my workstation. Unfortunately I just got timeouts. 😢 ![image](https://private-user-images.githubusercontent.com/8978655/296797936-953a0122-33cb-4850-9f70-b996e9274af4.png?

I'm still seeing the same as Peter unfortunately 😭

yonikid15 commented:
@esizer Okay so looks like the timeout changes and moving the `toBeVisible()` check fixed all the tests.
As requested, I attached a snapshot of all the tests and the duration for each browser. Here are the longest tests that would benefit from more fine grained timeout increases vs just a global increase:

1. Can submit application (~4min)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/64621732-2dd4-4d2e-bcc1-85468e4776bf)

2. Skill library (~30sec)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/fec60204-3b88-4183-bb8e-06ffe71812d6)

3. Cannot access restricted paths (admin ~50sec + applicant ~50sec)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/ca51205e-43fa-4750-9293-c1fb4cb31950)

4. Can logout (~1min)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/45f9446e-2532-477c-bfa4-61d3191118ce)

5. Redirects restricted pages to sign in (~1min)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/cf9beab6-bbd1-41d6-acb2-3288f640a111)

yonikid15 commented:
Here are list of the assertions that timeout:

## Authenticated:
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/4d0883b6-ed9c-4759-bc11-8498f7ebce72)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/30e3b3f5-dae9-4e36-843c-1f529345ab33)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/874a962e-b0d6-47e6-b060-0fa2220d003e)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/4d8f12c4-7ab2-4cc7-a547-7aa6fb0d5137)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/4c14889a-9b9a-4849-ba3b-44d1fa86bd7c)

## Skill Library

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/d6beb9ee-8564-4d37-87ab-1144416ef464)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/75c20236-65d7-43a0-b886-6c0fdfab27bf)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/838f7064-5fb1-4bbd-aca2-a463318a6aa2)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/8b62481a-9bc3-4e44-ae69-d8fc0c835bf0)

## Application

All browsers fail at same point (`Cannot read properties of null (reading 'id')` at u.id)
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/22059495/54f4de25-f546-46b2-be4b-091cb88dea7f)

vd1992 commented:
Trying it out
Running `npm run e2e:playwright`

Results: 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/62d3fbcb-e366-4409-aaeb-6b45b366b6dd)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/44e3c5a5-5f4e-4ae5-ae43-17eeee884641)

esizer commented:
> Trying it out

That one seems to be very troublesome 😛 There may be something wrong with the logic in it causing contexts to be closed too early 🤔 .

vd1992 commented:
Trying the UI bit and it seems to work? Two tests appear but a little hard to follow what happened. Appears to flash through and you have to click on the frames in the timeline to see what was rendered? 

> > Trying it out
> 
> That one seems to be very troublesome 😛 There may be something wrong with the logic in it causing contexts to be closed too early 🤔 .

Reran and got basically an identical result with a slight runtime difference
Expanded the sections out

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/e7ee7d95-d1bf-4fcb-b0d9-16ae085a69fc)

esizer commented:
> Here are list of the assertions that timeout:

Okay, I updated the timeout to be a little more reasonable (30 seconds). As well, I updated the restricted paths tests to wait for the URL instead of network idle. Hopefully this helps 🤷‍♀️ 

 - [tweak timeout to use lower value](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8917/commits/463429acad31d9f2b60513943d9429812c372508)
 - [improve waits for auth tests](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8917/commits/5dc9c85b9dfe6dc90a15cef5e101a53edbf5d623)

vd1992 commented:
Got 5-6 tests failing now

Last run 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/5b00742a-3e57-4c93-838b-80967c036f32)

Application submit all fail for the same reason

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/9c4baa2e-ec79-46ff-a19c-d5d4cbcb3042)

And the auth fails are

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/2bb38947-8ca2-4511-90b1-b1c8a6082af1)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/29f9343a-d315-4ce8-bfaf-8420c0bea11f)

vd1992 commented:
For this run, noticed this little problem with it not failing correctly

https://github.com/GCTC-NTGC/gc-digital-talent/actions/runs/7644060225/job/20827486364

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/fcdb1173-b5be-46d3-befe-96700b25138b)

esizer commented:
> And the auth fails are

Do your pages typically take a long time to load on your local machine? That was the case with @yonikid15. Setting the timeout to 30s caused timeouts. I just bumped it to 60s ([increase timeout to 60 seconds](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8917/commits/a1b55d95dd7fc5b4256d06bbf4ce5ed5408d95bf)) which got them to consistently pass for him.

> For this run, noticed this little problem with it not failing correctly

Yeah, I noticed that. Trying to get it to fail properly in the CI right now 🤞

vd1992 commented:
> > And the auth fails are
> 
> Do your pages typically take a long time to load on your local machine? That was the case with @yonikid15. Setting the timeout to 30s caused timeouts. I just bumped it to 60s ([increase timeout to 60 seconds](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8917/commits/a1b55d95dd7fc5b4256d06bbf4ce5ed5408d95bf)) which got them to consistently pass for him.
> 
> > For this run, noticed this little problem with it not failing correctly
> 
> Yeah, I noticed that. Trying to get it to fail properly in the CI right now 🤞

Not that long a load time locally. 
Reran, other than the Typescript error, this one persists

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/9a2b83bd-456d-4cc5-a654-ace8bf798600)

esizer commented:
> For this run, noticed this little problem with it not failing correctly

Got it to fail properly 😆 https://github.com/GCTC-NTGC/gc-digital-talent/actions/runs/7645913027/job/20833569267?pr=8917#step:9:241

> Not that long a load time locally.
> Reran, other than the TypeError, this one persists

Would you be willing to pair up so I can watch your screen and try and debug this problem in real time? I'm having trouble reproducing/debugging it 😢

vd1992 commented:
Running that one test by itself in UI mode and it appears to succeed?

vd1992 commented:
Running it a few times for each browser, that was a handy addition, chromium and firefox work. Webkit fails on "Can access restricted paths" only, but this test passes when I run it singled out in UI mode.  

Just putting that there as a point of comparison for others trying it out.

mnigh reviewed (approved):
❤️ this. 👋 🇨🇾!

tristan-orourke commented on apps/playwright/fixtures/AppPage.ts:
```diff
@@ -0,0 +1,86 @@
+   * @returns
+   */
+  async graphqlRequest(query: string, variables?: Record<string, unknown>) {
+    await this.gotoHome();
```
Why is this part necessary? I would think this step would just slow down graphql requests. If its to make sure the authoirzationQuery has run, would it be possible to try to get auth tokens, and if they don't exist yet, _then_ goToHome?

esizer replied:
This is related to the other comment...sort of. So, our tokens were expiring before the test ended and this request was actually happening outside of the browser. So this just goes home, waits for the token to be refreshed and then sends off the request. I think we can make this better in one of three ways:

1. As mentioned in other comment, intercept refreshes in the in the fixture and keep our storage state up to date (probably best?)
2. Right a refresh to use here and in the other graphql utility to refresh if the token is expired. Basically, mimic what our client provider is doing
3. Just be lazy and extend our dev tokens to something like 30mins so this is unnecessary 😅 

If you'd like, I could see about getting one of these solutions working here before merging.

tristan-orourke commented on apps/playwright/tests/admin/auth.spec.ts:
```diff
@@ -0,0 +1,164 @@
+
+test.describe("Authenticated", () => {
+  test("Has tokens", async ({ applicantPage }) => {
+    await applicantPage.gotoHome();
```
So if you use the applicantPage fixture, it always comes authenticated? That's cool! Where does it actually happen? `fixtures/ApplicantPage.ts` doesn't look like it does anything...

esizer replied:
Yeah! I really liked this. We can also do other roles too! It happens in the `global.setup.ts` file before everything runs.

We log the user in then it takes a snapshot of the current app state (cookies, etc.) and stores it in a JSON file under `.auth`. Then when you use this fixture, it loads that into the browser. As mentioned before, it gets stale during testing since this seem to only happen once. I offered some solutions to attempt to keep this file up to date though.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/95c8c0d584e44f6f96dba759e8c6388b6137fd1d/apps/playwright/global.setup.ts#L6-L14

tristan-orourke reviewed (commented):
Overall, the ways this is set up, with fixtures and util functions, makes a lot of sense to me! I did leave some questions about things I'd like to understand better.

esizer commented on apps/playwright/utils/graphql.ts:
```diff
@@ -0,0 +1,42 @@
+  variables?: Record<string, unknown>,
+  storageState?: string,
+) {
+  const browser = await chromium.launch();
```
Yeah, that is a good thing to bring up. Originally, that was how things were happening and I really do agree that is how we should be doing it. This was an easy fix to get the auto refresh of our client since the access token is short lived and usually expires before the test is done. 

I should have written something about this so we could fix it but this was the easy fix. I think we could write something into the fixture to intercept refresh requests and update the auth storage state so it is always up to date. There are other ways but this is likely the best solution.

esizer commented on .github/workflows/playwright.yml:
```diff
@@ -0,0 +1,100 @@
+      - name: Install Playwright Browsers
+        run: npx playwright install --with-deps
+
+      - name: Run Chromium Tests
+        run: npm run e2e:playwright:chromium
+      - uses: actions/upload-artifact@v3
+        if: always()
+        with:
+          name: chromium-report
+          path: apps/playwright/playwright-report/
+          retention-days: 30
+
+      - name: Run Firefox Tests
+        run: npm run e2e:playwright:firefox
+      - uses: actions/upload-artifact@v3
+        if: always()
+        with:
+          name: firefox-report
+          path: apps/playwright/playwright-report/
+          retention-days: 30
+
+      - name: Run Webkit Tests
```
It does not, you can do that though by [passing the workers flag and setting it to 1](https://playwright.dev/docs/test-parallel#disable-parallelism).

I added the browser specific commands so we could still run all the tests but reduce the run time significantly.

tristan-orourke commented:
Decide whether to show errors in files changed tab
Show report on error
Remove the overlapping cypress test

mnigh commented on apps/playwright/playwright.config.ts:
```diff
@@ -0,0 +1,90 @@
+  reporter: process.env.CI
+    ? [["github"], ["dot"]]
+    : [["line"], ["html", { open: "on-failure" }]],
+  timeout: 5 * 60 * 1000, // 5 mins
```
Can we removed this?

mnigh commented on apps/playwright/playwright.config.ts:
```diff
@@ -0,0 +1,90 @@
+    ? [["github"], ["dot"]]
+    : [["line"], ["html", { open: "on-failure" }]],
+  timeout: 5 * 60 * 1000, // 5 mins
+  expect: { timeout: 3 * 60 * 1000 }, // 3 mins
```
Can we removed this?

tristan-orourke reviewed (approved):
🎉 🎉 🎉
